openfermioncirq/primitives/general_swap_network.py

```diff
@@ -0,0 +1,117 @@
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+from typing import Dict, List, Sequence, Tuple
+
+import cirq
+import cirq.contrib.acquaintance as cca
+import numpy as np
+import openfermion
+import scipy.linalg as la
+
+
+import openfermioncirq.gates as ofc_gates
+
+
+class GreedyExecutionStrategy(cca.GreedyExecutionStrategy):
+    def get_operations(self,
+            indices: Sequence[int],
+            qubits: Sequence[cirq.Qid],
+            ) -> cirq.OP_TREE:
+        index_set = frozenset(indices)
+        if index_set not in self.index_set_to_gates:
+            return ()
+
+        n_indices = len(index_set)
+        abs_to_rel = dict(zip(indices, range(n_indices)))
+        gates = self.index_set_to_gates.pop(index_set)
+        for gate_indices, gate in sorted(gates.items()):
+            if len(index_set) == 1:
+                yield gate(*qubits)
+            elif isinstance(gate, ofc_gates.FermionicSimulationGate):
+                pos = [abs_to_rel[i] for i in gate_indices]
+                yield gate.permuted(pos)(*qubits)
+            else:
+                perm = dict(zip((abs_to_rel[i] for i in gate_indices),
+                    (abs_to_rel[j] for j in indices)))
+                reverse_perm = dict((j, i) for i, j in perm.items())
+                yield cca.LinearPermutationGate(n_indices, perm,
+                    ofc_gates.FSWAP)(*qubits)
+                yield gate(*qubits)
+                yield cca.LinearPermutationGate(n_indices, reverse_perm,
+                    ofc_gates.FSWAP)(*qubits)
+
+
+def trotter_circuit(
+        swap_network: cirq.Circuit,
+        initial_mapping: Dict[cirq.LineQubit, int],
+        hamiltonian: openfermion.InteractionOperator,
+        execution_strategy: cca.executor.ExecutionStrategy =
+            GreedyExecutionStrategy,
+        ) -> cirq.Circuit:
+    """Uses a swap network to construct a circuit that approximates exp(i H)
+    for the specified H.
+
+    Specifically, for each (unordered) set of modes $I = \{i_1, \ldots, i_k\}$,
+    there is at most one gate $\exp(i H_I)$ corresponding to the term  $H_I$ in
+    the Hamiltonian that involves exactly those modes.
+
+    Args:
+        swap_network: The circuit containing permutation gates and acquaintance
+            opportunity gates.
```

Comment:
Explain in more detail what this circuit is supposed to be i.e., what the swap gates in it do and how one might go about constructing it.

Reply:
Done